### PR TITLE
GEOT-6637  Upgrade to JTS Topology Suite 1.17.0

### PR DESCRIPTION
**Pull request by jnh5y**
Signed-off-by: Jim Hughes <jnh5y@ccri.com>

The goal of this PR is to upgrade to the latest release of JTS, version 1.17.0. 

At present, this PR is a WIP until there's a release of JAI-EXT and JAITOOLS which leverage the new version.  (This is due to a small API change involving a method switching between LineStrings and LinearRings.)

Depending on:

* [x] https://github.com/geosolutions-it/jai-ext/pull/269: now released as jaiext-1.1.16 and deployed to geosolutions repo
* [x] https://github.com/bjornharrtell/flatgeobuf/pull/65: now tagged as flatgeobuf 3.4.0 and deployed to maven central
* [x] https://github.com/ElectronicChartCentre/java-vector-tile/pull/32: now released as java-vector-tile 1.3.9 (and deployed to repo.osgeo.org)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [ ] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by jnh5y** [comment]
This PR requires https://github.com/geosolutions-it/jai-ext/pull/269 to be merged and a release of JAI-Ext to be made.  After that this PR can be updated.  It, https://github.com/GeoWebCache/geowebcache/pull/868 and https://github.com/geoserver/geoserver/pull/4372 will cover upgrading GT/GWC/GS to use the latest JTS.

**Comment by jnh5y** [comment]
This PR (and the downstream GWC/GS ones) depends on an update in and a release of Jai-ext and flatgeobuf.  The latter is tracked here: https://github.com/bjornharrtell/flatgeobuf/pull/65.

**Comment by jnh5y** [comment]
GeoServer may require an update to java-vector-tile.  I have a pr here: https://github.com/ElectronicChartCentre/java-vector-tile/pull/32.

**Comment by aaime** [comment]
Seeing a new flatgeobuf release: https://github.com/bjornharrtell/flatgeobuf/releases/tag/3.4.0

**Comment by jnh5y** [comment]
Awesome!  I see that java-vector-tile released as well.  I'll update the PRs and re-check all the builds today!

**Comment by jnh5y** [comment]
I'm not seeing java-vector-tile on a Maven repository yet.  As soon as that happens, this should be good to go!

**Comment by jodygarnett** [comment]
Okay Java 11 is failing on:
```
[INFO] -------------------------------------------------------------
[WARNING] COMPILATION WARNING : 
[INFO] -------------------------------------------------------------
[WARNING] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java:[176,21] reverse() in org.locationtech.jts.geom.LinearRing has been deprecated
[WARNING] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CircularString.java:[215,21] reverse() in org.locationtech.jts.geom.LineString has been deprecated
[WARNING] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CompoundCurve.java:[190,21] reverse() in org.locationtech.jts.geom.LineString has been deprecated
[WARNING] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CompoundCurve.java:[194,50] reverse() in org.locationtech.jts.geom.LineString has been deprecated
[WARNING] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CompoundRing.java:[116,21] reverse() in org.locationtech.jts.geom.LinearRing has been deprecated
[INFO] 5 warnings 
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/geotools/geotools/modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java: warnings found and -Werror specified
[INFO] 1 error
```

**Comment by aaime** [comment]
Odd failure in Travis JDK11... don't know how it can relate to JTS though. Mentioning it here and restarting the build:

```
[ERROR] Tests run: 18, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.044 s <<< FAILURE! - in org.geotools.filter.text.ecql.ECQLTemporalPredicateTest
[ERROR] filterWithOgcInstantToEcql(org.geotools.filter.text.ecql.ECQLTemporalPredicateTest)  Time elapsed: 0.008 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<... 2020-07-06T23:54:02[.000]+00:00> but was:<... 2020-07-06T23:54:02[]+00:00>
	at org.geotools.filter.text.ecql.ECQLTemporalPredicateTest.filterWithOgcInstantToEcql(ECQLTemporalPredicateTest.java:141)
```

**Comment by aaime** [comment]
Odd failure in Travis JDK11... don't know how it can relate to JTS though. Mentioning it here and restarting the build:

```
[ERROR] Tests run: 18, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.044 s <<< FAILURE! - in org.geotools.filter.text.ecql.ECQLTemporalPredicateTest
[ERROR] filterWithOgcInstantToEcql(org.geotools.filter.text.ecql.ECQLTemporalPredicateTest)  Time elapsed: 0.008 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<... 2020-07-06T23:54:02[.000]+00:00> but was:<... 2020-07-06T23:54:02[]+00:00>
	at org.geotools.filter.text.ecql.ECQLTemporalPredicateTest.filterWithOgcInstantToEcql(ECQLTemporalPredicateTest.java:141)
```

Also legit failure in the QA build, mis-formatted file, sign of a code change without a sub-sequent local maven build (mvn install reformats the files for you):

```
[ERROR] Found 1 non-complying files, failing build
[ERROR] To fix formatting errors, run "mvn fmt:format"
[ERROR] Non complying file: /home/travis/build/geotools/geotools/modules/unsupported/process-feature/src/main/java/org/geotools/process/vector/ClipProcess.java
```

File: modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java [hunk 1]
```diff
@@ -173,7 +173,7 @@ public CircularRing normalizeRing() {
         return new CircularRing(ncp, factory, delegate.getTolerance());
     }
 
-    public Geometry reverse() {
+    public Geometry reverseInternal() {
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java:176]: @aaime this is the "functional" change for the PR.

The superclass now wants to do some book keeping and call reverseInternal() to do the work.

Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java:176]: reverseInternal is public? Really?

Comment (jodygarnett) [reply modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java:176]: probably protected, let me look:  [reverseInternal()](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geom/Geometry.java#L1303)
```
protected abstract Geometry reverseInternal();
```

Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/geometry/jts/CircularRing.java:176]: I'd keep it that way then :D

**Comment by jnh5y** [comment]
Ok, my side branch (with the GH pointing at the other branches) built successfully https://github.com/geotools/geotools/pull/3052.  I did have to futz with the GH actions for a bunch of commits.

**Comment by jnh5y** [comment]
I think this is good to merge.  @jodygarnett and @dr-jts talked about the JTS api bits.  It looks like LinearRing has a public reverse method which makes some of the current approach a compromise.

Any more feedback?

**Comment by jodygarnett** [comment]
We are ready to go, remember to remove https://github.com/geotools/geotools/pull/3052 when this is done

**Pull request merged**